astropy/units/quantity_helper/function_helpers.py

```diff
@@ -33,22 +33,56 @@
 return a Quantity directly using ``quantity_result, None, None``.
 
 """
-
 import functools
 import operator
+import warnings
 
 import numpy as np
 from numpy.lib import recfunctions as rfn
 
-from astropy.units.core import dimensionless_unscaled
+from astropy.units import (
+    Quantity,
+    LogQuantity,
+    StructuredUnit,
+    dimensionless_unscaled,
+    dex,
+    mag,
+    dB,
+)
 from astropy.units.errors import UnitConversionError, UnitsError, UnitTypeError
 from astropy.utils.compat import (
     COPY_IF_NEEDED,
+    NUMPY_LT_1_24,
     NUMPY_LT_2_0,
     NUMPY_LT_2_1,
     NUMPY_LT_2_2,
     NUMPY_LT_2_4,
 )
+from astropy.utils.exceptions import AstropyUserWarning
+
+# INTERNAL IMPORTS - CRITICAL FOR REGISTRATION
+from .helpers import function_helper, dispatched_function
+from .util import _as_quantity, _as_quantities, _quantities2arrays
+
+def _get_np_func_name(func):
+    """
+    Safely get the name of a numpy function.
+    If the function is already a string, return it.
+    """
+    if isinstance(func, str):
+        return func
+    return getattr(func, "__name__", str(func))
```

Comment:
I think it would be preferable to only use static lookups (i.e. `func.__name__`) and only allow one type (functions) as inputs. This will make callsite much easier to reason about.
Additionally, it might not be sufficient to retrieve the name of a function, as we sometimes also need to carry the name of it's parent(s) namespaces. Not sure how this applies yet, but that's something to keep in mind.